relay-server/src/actors/processor.rs

```diff
@@ -936,68 +936,295 @@ impl EnvelopeProcessorService {
-            &self.inner.config,
-            self.inner.outcome_aggregator.clone(),
+        relay_log::trace!(
+            "Items in the incoming envelope: {:?}",
+            state
+                .envelope()
+                .items()
+                .map(|item| item.ty())
+                .collect::<Vec<_>>()
         );
-        replay::process(state, &self.inner.config)?;
-        profile::filter(state);
-        span::filter(state);
 
-        if state.creates_event() {
-            // Some envelopes only create events in processing relays; for example, unreal events.
-            // This makes it possible to get in this code block while not really having an event in
-            // the envelope.
+        let Some(ty) = state
+            .event_type()
+            // get the item type from the event type if there is an event
+            .map(ItemType::from_event_type)
+            // othrwise get the first item from the items list and assume it's the only one defined
+            // which we are waiting for.
+            .or_else(|| {
+                state
+                    .envelope()
+                    .items()
+                    .next()
+                    .map(|item| item.ty().clone())
+            })
+        else {
+            relay_log::error!("There are not type and no items in the envelope");
+            return Err(ProcessingError::NoEventPayload);
+        };
 
-            if_processing!({
-                unreal::expand(state, &self.inner.config)?;
-            });
+        // Defined processing pipeilnes based on the event type and/or item type they contain.
+        match ty {
+            // This stil can contain attachements.
+            ItemType::Event => {
+                event::extract(state, &self.inner.config)?;
 
-            event::extract(state, &self.inner.config)?;
-            profile::transfer_id(state);
+                if_processing!({
+                    attachment::create_placeholders(state);
+                });
 
-            if_processing!({
-                unreal::process(state)?;
-                attachment::create_placeholders(state);
-            });
+                event::finalize(state, &self.inner.config)?;
+                self.light_normalize_event(state)?;
+                dynamic_sampling::normalize(state);
+                event::filter(state)?;
+                dynamic_sampling::run(state, &self.inner.config);
+                dynamic_sampling::sample_envelope(state)?;
 
-            event::finalize(state, &self.inner.config)?;
-            self.light_normalize_event(state)?;
-            dynamic_sampling::normalize(state);
-            event::filter(state)?;
-            dynamic_sampling::run(state, &self.inner.config);
+                if_processing!({
+                    event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+                    self.enforce_quotas(state)?;
+                });
+
+                if state.has_event() {
```

Comment:
Always true?

Reply:
Sooo, it's tricky, the `enforce_quotas`  can actually remove the event from the processing state 😄  and this condition will be false. And same goes for every question you asked where we enforce quotas.

Reply:
Thanks!